pilot/pkg/networking/plugin/authz/rbac.go

```diff
@@ -0,0 +1,387 @@
+	"istio.io/istio/pilot/pkg/serviceregistry/kube"
+	"istio.io/istio/pkg/log"
+)
+
+const (
+	// RbacFilterName is the name for the Rbac filter.
+	// TODO(yangminzhu): Update once the final name is decided.
+	RbacFilterName = "envoy.filters.http.rbac"
+
+	sourceIP        = "source.ip"
+	sourceService   = "source.service"
+	destinationIP   = "destination.ip"
+	destinationPort = "destination.port"
+	userHeader      = ":user"
+	methodHeader    = ":method"
+	pathHeader      = ":path"
+	//TODO(yangminzhu): Update the key after the final header name of service is decided.
+	serviceHeader       = ":service"
+	requestHeaderPrefix = "request.header["
+	requestHeaderSuffix = "]"
+)
+
+// Plugin implements Istio RBAC authz
+type Plugin struct{}
+
+// NewPlugin returns an instance of the authz plugin
+func NewPlugin() plugin.Plugin {
+	return Plugin{}
+}
+
+// OnOutboundListener is called whenever a new outbound listener is added to the LDS output for a given service
+// Can be used to add additional filters on the outbound path
+func (Plugin) OnOutboundListener(in *plugin.InputParams, mutable *plugin.MutableObjects) error {
+	return nil
+}
+
+// OnInboundListener is called whenever a new listener is added to the LDS output for a given service
+// Can be used to add additional filters (e.g., mixer filter) or add more stuff to the HTTP connection manager
+// on the inbound path
+func (Plugin) OnInboundListener(in *plugin.InputParams, mutable *plugin.MutableObjects) error {
+	enabled, err := isRbacEnabled(in.Env.IstioConfigStore)
+	if err != nil {
+		return fmt.Errorf("rbac plugin failed to enable: %v", err)
+	}
+	// Only supports sidecar proxy of HTTP listener for now.
+	if !enabled || in.Node.Type != model.Sidecar || in.ListenerType != plugin.ListenerTypeHTTP {
+		return nil
+	}
+
+	filter, err := buildHTTPFilter(in.ServiceInstance.Service.Hostname, in.Env.IstioConfigStore)
+	if err != nil {
+		return err
+	}
+	for _, chain := range mutable.FilterChains {
+		chain.HTTP = append(chain.HTTP, filter)
+	}
+
+	return nil
+}
+
+// OnInboundCluster implements the Plugin interface method.
+func (Plugin) OnInboundCluster(env model.Environment, node model.Proxy, service *model.Service,
+	servicePort *model.Port, cluster *xdsapi.Cluster) {
+}
+
+// OnOutboundRouteConfiguration implements the Plugin interface method.
+func (Plugin) OnOutboundRouteConfiguration(in *plugin.InputParams, route *xdsapi.RouteConfiguration) {
+}
+
+// OnInboundRouteConfiguration implements the Plugin interface method.
+func (Plugin) OnInboundRouteConfiguration(in *plugin.InputParams, route *xdsapi.RouteConfiguration) {
+}
+
+// OnOutboundCluster implements the Plugin interface method.
+func (Plugin) OnOutboundCluster(env model.Environment, node model.Proxy, service *model.Service,
+	servicePort *model.Port, cluster *xdsapi.Cluster) {
+}
+
+func isRbacEnabled(store model.IstioConfigStore) (bool, error) {
+	rbacConfigs, err := store.List(model.RbacConfig.Type, kube.IstioNamespace)
```

Comment:
I know this pattern is used everywhere in the code - however for 1.0 we need to change it for existing code, and stop using it for new code.

This is called for each listener of each pod - and internally converts all yaml to json to proto, every single time. Just to find if there is one entry. Same later - when it iterates each list. 

Can we use some better data structures ? Maybe some caching ?

Reply:
By adding cache, do you mean to do so outside the scope of the rbac plugin? Like we call store.List() only once and generate the intermediate data and make the plugin to read from the intermediate data. 
One approach I can think of is to provide a initialize function to the [plugin object](https://github.com/istio/istio/blob/ec41ba682450c412c7d19174ae7a21c52f554844/pilot/pkg/networking/plugin/plugin.go#L91) and call it (probably) inside [buildSidecarInboundListeners()](https://github.com/istio/istio/blob/ec41ba682450c412c7d19174ae7a21c52f554844/pilot/pkg/networking/core/v1alpha3/listener.go#L230), So that the conversion from yaml to proto only happens once for each node, instead of for each single listener?

(I feel it doesn't make sense to use any kind of cache inside the plugin, we don't have a good mechanism to invalidate the out-dated cached data inside the plugin, do we?)

I would prefer to do this in a different PR if you think the approach is okay. Thank you!